src/docs/arch-block.dox

```diff
@@ -1,9 +1,195 @@
+read the block at the location specified and return the page image as a buffer
+to the layer above. This will then be instantiated as a page in memory.
+
+From there subsequent page addresses can be read from the root page and the
+process repeated as required. If a cursor traverses to a page which hasn't been
+read into memory the same process will take place.
+
+@subsection Writing
+
+Two cases exist for writing out data using the block manager, checkpoint and
+eviction. Both are explained in the following two sections. When a page image
+is written out the block manager the \c bm->write API is called. See \c bt_io.c
+for more detail.
+
+@subsubsection Checkpoint
+
+For details on checkpoint at the WiredTiger level see @ref arch-checkpoint.
+
+At the block manager level a checkpoint corresponds with a set of blocks which
+are stored in the file associated with the given URI. Typically each file will
+contain a minimum of two checkpoints. When opening an existing file the latest
+checkpoint is read.
+
+When WiredTiger performs a checkpoint clean pages will not result in additional
+blocks being created and as such existing blocks from older checkpoints may be
+reused, this means that while there may be more than one checkpoint in a file
+the block are not guaranteed to be mutually exclusive. Checkpoints are created
+in depth first order, leaf blocks are created, then the parent blocks. This is
+a requirement as the parent blocks contain the addresses of the leaf blocks.
+
+The block manager doesn't guarantee that calling \c bm->write will result in
+the data being flushed to disk. In the checkpoint scenario WiredTiger will also
+call \c bm->sync which will call the file system dependent flush function.
+
+As a checkpoint progresses it updates its three extent lists, these extent lists
+are written out to disk as part of the checkpoint in blocks. When a checkpoint
+is written the previous checkpoints will be cleaned up if possible, suppose we
+have finished writing out checkpoint C and have checkpoints A and B on disk.
+We only need to retain two checkpoints so we can cleanup checkpoint A.
+Checkpoint A's extent lists get merged into checkpoint B's extent lists. If
+a block, N, is in the alloc list of checkpoint A and the removed discard list
+of checkpoint B the block is considered free. Until a block in an alloc list
+is seen in a discard list the space cannot freed.
+
+@subsubsection Eviction
+
+Eviction also utilizes the block manager, when a page is evicted and contains
+data that needs to be maintained, logically a block needs to be written.
+Eviction also calls \c bm->write however it does not instruct the block manager
+to sync the data.
+
+@subsection Compaction
+
+As WiredTiger maintains two checkpoints by default the file size will often be a
+multiple of the data size modulo the compaction ratio. And given that new blocks
+are written by default where they fit best it's common that removal of data will
```

Comment:
_And given that new blocks are written by default where they fit best it's common that removal of data will not result in the file shrinking. As the file can only be shrunk when there are available blocks at the end of the file._

Maybe reword similar to this:

_By default, the block manager writes new blocks where they fit best. This means that blocks can be written at any location in the file. As the file can only be shrunk when there are available blocks at the end of the file, there is no guarantee that removing data will reduce the file size._

Reply:
I wasn't a huge fan of that re-wording but I've reworded some of it in: 477de51a3